### ISSUE TITLE
Hass-io components throw 'NewConnectionError' when pi-hole is enabled

### ISSUE THREAD

**Issue by GeertvanHorrik** [hassio-addons/addon-pi-hole#63]
# Problem/Motivation

Getting lots and lots of 'NewConnectionError' errors on Telegram, Ring and more. After some investigation, there are more users experiencing this. Whenever I turned off pi-hole, everything started working immediately.

## Expected behavior

Connections to be able to set up.

## Actual behavior

Lots of errors in the log:

```
Platform rest not ready yet. Retrying in 90 seconds. 
11:00 AM helpers/entity_platform.py (WARNING) 
Error fetching data: <PreparedRequest [GET]> from http://ip.jsontest.com/ failed with HTTPConnectionPool(host='ip.jsontest.com', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x6cc7a130>: Failed to establish a new connection: [Errno -3] Try again',)) 
11:00 AM components/sensor/rest.py (ERROR) 
Platform rest not ready yet. Retrying in 60 seconds. 
10:59 AM helpers/entity_platform.py (WARNING) 
Error fetching data: <PreparedRequest [GET]> from http://ip.jsontest.com/ failed with HTTPConnectionPool(host='ip.jsontest.com', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x6cc64430>: Failed to establish a new connection: [Errno -3] Try again',)) 
10:59 AM components/sensor/rest.py (ERROR) 
```

Same for ring, telegram, etc.

I double checked, the dns entries are not even being requested in pi-hole itself.

## Steps to reproduce

1. Install hass.io
2. Install the pi-hole add-on
3. Update router to use pi-hole as dns server

## Proposed changes

Not sure how to fix this. If this is a configuration error, I would like to know and I can update the docs.

**Comment by GeertvanHorrik** [comment]
The hass-io already had a static ip (otherwhise I couldn't point the router to pi-hole). I couldn't find the note about the external DNS servers in the docs though. Will check that, thanks!

**Comment by Ornias1993** [comment]
I know your issue and franck is both right and wrong here:
There are three router config options:
1. Your router is using pihole as its dns server and runs a dns locally
2. Your router DHCP points all devices to the Pihole DNS server
3. Your router forces all devices to use the Pihole DNS server

1. if your hassio system points to the router, it will ask pihole which would ask externally, this is not efficient in terms of latency. But would not lead to a loop.
2. In the current build of PiHole the dns server it recieves by dhcp gets overruled for piholed dns queries.
3. This would lead to a loop condition, you either need to list your pihole/hassio box "do not force redirect dns" (not many routers have that feature) or point Pihole to a DNS-Https proxy (to bypass dns all together and thus get past the forced router redirect).

- Static IP is only required when using pihole DHCP, otherwise dhcp-reserved or even dhcp ip works perfectly fine.
- I don't know if pihole is loaded fast enough on boot, if it doesn't you may indeed need an external dns server for hassio. If it does that would solve nothing.

**Comment by GeertvanHorrik** [comment]
@Ornias1993 thanks for your answer. I was using DHCP (it would get a fixed ip address all the time), but then I read that "DHCP is not the same as a static ip" comment by @frenck (see https://github.com/home-assistant/hassio/issues/834#issuecomment-443680740). I couldn't think of any reason why it would, but I assumed he had his reasons for saying this.

In my understanding, the real only problem is hassio getting pi-hole as DNS server, but I have a secondary DNS set up in my router (UniFi), so it should still be able to get passed this when pi-hole is still starting up (right?!):

![image](https://user-images.githubusercontent.com/1246444/51959583-59449e00-2455-11e9-8fa9-7c916cd0302e.png)

I am using:

1. A router with a DHCP reservation specifically for hassio (and thus pi-hole)
2. The router is configured to pass pi-hole as primary dns, but also provides a secondary in case pi-hole is down for whatever reason

**Comment by Ornias1993** [comment]
I don't know for sure how it would work with the current unspecified IP blocking pihole employs, but putting in a secondary may cause blocked traffic to circumvent the pihole, did you test this setup?

But anyhow, my best guess would be that some addons don't execute a query on the secondary DNS server if the first query fails. Secondary DNS is nice to have, but if the primary fails not every program/us/config could handle it.

In case you do use a secondary, you should use the external primary dns server, not their failover, just for claimnesssake. (it should be: local, 8.8.8.8, 8.8.4.4, if you want it that way)

You have two options:
1. Put a static DNS server (8.8.8.8 or 1.1.1.1) on the hassio box (and remove the odd 8.8.4.4 secondary)
2. Keep your pihole outside hassio on a seperate box for now.

I have opted for option 2 for the time being.
- That way I could keep core network devices on a seperate vlan, could more easily add cloudflared and don't have a single point of failure (aka the internet still works if hassbox goes down).
- This also gives me the option of having a forced DNS redirect from the router to the pihole, which would not be possible on hassio-pihole due to the hassio-pihole issues. This filters out a lot of (microsoft) junk that tries to circumvent pihole.

TL:DR seperate box currently has more features and usecases than hassio-pihole, but if you want to go hassio-pihole, make sure the DNS on the hassio/hassos box is fixed.

**Comment by GeertvanHorrik** [comment]
I've installed pi-hole on my NAS using docker, that works great and keeps connections on HA alive.

**Comment by frenck** [comment]
@GeertvanHorrik 

> "A router with a DHCP reservation specifically for hassio"

That is not a static IP. That is a DHCP reservation, which is not the same.
